[FEATURE] Afficher Pix Orga dans la langue de l'utilisateur enregistrée dans son compte (PIX-5766)

lego-technix opened the pull request:
## :unicorn: Problème

Quand un utilisateur se connecte sur orga.pix.org le site web n'est pas dans la langue que l'utilisateur a configurée dans son compte. On souhaite que l'application Pix Orga soit traduite dans la langue que l'utilisateur a configurée dans son compte.

On rappelle ci-dessous la règle de priorité de la lang/locale des différents éléments.

**Règle de priorité** : `lang` utilisateur < `lang` query param < `lang` domaine `.fr`

## :robot: Proposition

L'information de la `lang` du user est déjà envoyée par l'API lors de l'authentification, il suffit de l'utiliser en respectant la règle de priorité. 

## :rainbow: Remarques

### Comportement

Le comportement actuel de Pix Orga est de changer la `lang` de l'utilisateur si un query param `lang` est passé dans l'URL. Ce n'est pas le comportement qu'on veut à terme car :
* c'est fait sans demander son consentement à l'utilisateur
* tout changement, de langue ou de toute autre propriété/préférence de l'utilisateur, par le simple fait de visiter un URL (qui est donc une requête GET) ayant certains query params (ici lang) est une mauvaise pratique. Sources :
   * https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET
   * https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Request_methods

Néanmoins, ce comportement du patch de l'utilisateur, qui avait du sens avant cette PR, n'est pas modifié dans cette PR-ci, de manière à donner le temps à tous les utilisateurs, responsables du support, etc. de s'approprier la nouvelle fonctionnalité apportée par cette PR, à savoir afficher Pix Orga dans la langue de l'utilisateur enregistrée dans son compte.

### Implémentation

L'implémentation proposée est basée sur le fait que tous les utilisateurs ont obligatoirement la colonne `lang` remplie dans la table `users` comme l'exige le schéma de la BDD. Faire une requête en production pour s'assurer que c'est bien le cas.

### Nommage dans le code

_TLD_ veut dire [Top-level domain](https://en.wikipedia.org/wiki/Top-level_domain), _domaine de premier niveau_ en français. Et donc la constante `FRANCE_TLD = 'fr'` correspond au _domaine de premier niveau_ pour la France.

### Refactor

Cette PR inclut des commits _refactor_ qui suppriment `dayjs.self.locale(locale)`, qui sont des appels inutiles, dans toutes les applis.

En effet le `setLocale` de ember-dayjs _set_ aussi la locale du service : https://github.com/sinankeskin/ember-dayjs/blob/f62de54409f21043f91de55724c63002c70180e6/addon/services/dayjs.js#L16-L20
dayjs.js
```javascript
  setLocale(locale) {
    this.useLocale(locale);

    this.locale = locale;
  }
```

## :100: Pour tester

1. Choisir un utilisateur ayant accès à Pix Orga, par exemple `allorga@example.net`
2. Se connecter sur Pix App **.org** https://app-pr5991.review.pix.org/
3. Aller dans « Mon compte »
4. Changer la langue de « Français » à « Anglais »
5. Aller sur Pix Orga **.fr** https://orga-pr5991.review.pix.fr/connexion puis avec le même URL mais en passant en plus le paramètre `?lang=en`  https://orga-pr5991.review.pix.fr/connexion?lang=en et constater que dans les deux cas le domaine est l'élément le plus prioritaire pour la sélection de la langue d'affichage, c'est à dire que le site web est toujours en français peu importe la présence du paramètre `lang`
6. Se connecter sur Pix Orga **.fr** https://orga-pr5991.review.pix.fr/ avec le même user que précédemment choisi et constater que le domaine est l'élément le plus prioritaire pour la sélection de la langue d'affichage, c'est à dire que le site web est toujours en français peu importe que l'utilisateur soit connecté ou non
7. Aller sur Pix Orga **.org**  https://orga-pr5991.review.pix.org/ sans être connecté et constater que le site web est en français par défaut
8. Se connecter sur Pix Orga **.org**  https://orga-pr5991.review.pix.org/ avec le même user que précédemment choisi et constater que le site web passe en anglais, qui est la `lang` de l'utilisateur
9. Aller sur Pix Orga **.org** https://orga-pr5991.review.pix.org/ avec l'utilisateur toujours connecté en faisant varier les valeurs du paramètre `lang`, en mettant alternativement `?lang=fr` https://orga-pr5991.review.pix.org/?lang=fr et `?lang=en` https://orga-pr5991.review.pix.org/?lang=en et constater que : 
   * le query param est pris en compte prioritairement à la langue de l'utilisateur
   * la `lang` de l'utilisateur a été modifiée en base cf. _Remarques > Comportement_ ci-dessus
11. Recharger la page de Pix Orga **.org**  https://orga-pr5991.review.pix.org/ avec l'utilisateur toujours connecté et constater que le site web s'affiche en anglais, qui est la `lang` de l'utilisateur, c'est à dire que le query param qui était « mémorisé » dans l'application web Pix Orga pour la précédente visite du site web est oublié et que c'est maintenant la `lang` de l'utilisateur qui est prise en compte
12. Vérifier qu'une valeur invalide pour `lang`, c'est à dire autre que `fr` ou `en`, ne fait jamais planter l'application Ember Pix Orga https://orga-pr5991.review.pix.fr/?lang=invalid-value  https://orga-pr5991.review.pix.org/?lang=invalid-value : ouvrir la console et constater qu'il n'y a pas d'erreur et on doit pouvoir naviguer normalement sans que certaines actions soient impossibles

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr5991.review.pix.fr)
  * [App (.org)](https://app-pr5991.review.pix.org)
  * [Orga](https://orga-pr5991.review.pix.fr)
  * [Certif](https://certif-pr5991.review.pix.fr)
  * [Admin](https://admin-pr5991.review.pix.fr)
  * [API](https://api-pr5991.review.pix.fr/api/)
  * [Pix1D](https://pix1d-pr5991.review.pix.fr)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr5991/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr5991/environment)

lego-technix committed:
refactor(orga): group and regroup tests contexts so the tests can be extended more easily

lego-technix committed:
feat(orga): add new getter isFranceDomain in current-domain service and leverage it instead of url.isFrenchDomainExtension

This way the currentDomain.isFranceDomain property is more logically placed than
url.isFrenchDomainExtension and Pix Orga an Pix Certif are more alike.

lego-technix committed:
feat(orga): when the user is authenticated set ember's locale according to logic: user's lang < lang query param < .fr domain lang

lego-technix committed:
refactor(certif): reorganize code and rename so that Pix Certif and Pix Orga are more alike

annemarie35 commented on orga/app/services/url.js:
```diff
@@ -2,7 +2,7 @@ import Service, { inject as service } from '@ember/service';
 import ENV from 'pix-orga/config/environment';
 
-const FRENCH_DOMAIN_EXTENSION = 'fr';
+const FRENCH_LANG = 'fr';
```
Attention, je crois que FRENCH_LANG n'existe pas du tout.
J'ai commencé à repertorier les constantes qu'on utilise histoire d'être consistant https://1024pix.atlassian.net/wiki/spaces/DEV/pages/3745480718/Ce+que+nous+avons+appris+en+codant

FRENCH_DOMAIN_EXTENSION est devenue FRANCE_TLD dans les tickets récent, je propose qu'on le garde

lego-technix replied:
J'ai défini cette constante `FRENCH_LANG` car on fait la comparaison avec ce qui « semble » être une « langue » qui est récupérée par `const currentLanguage = this.intl.t('current-lang')`. Mais le service `intl` n'est pas clair son API est parfois orientée langue et parfois orientée locale.

Donc tout à fait d'accord pour utiliser une constante orientée locale à la place de `const FRENCH_LANG = 'fr'`. Mais par contre ce n'est pas `FRENCH_DOMAIN_EXTENSION` ou `FRANCE_TLD` qu'il faut utiliser à la place car il faut comparer des choses comparables, on ne va pas comparer une langue ou locale avec un TLD.

lego-technix replied:
Je viens de renommer `FRENCH_LANG` en `FRENCH_LOCALE`.

Il y a dans `orga/app/adapters/application.js` un autre const `FRENCH_LOCALE = 'fr-fr'` qui n'est ni au format canonique, ni bien nommée. Mais c'est un autre sujet et qui n'a pas à être traité dans cette PR.

er-lim commented on orga/app/services/session.js:
```diff
@@ -25,35 +31,42 @@ export default class CurrentSessionService extends SessionService {
+  _setLocale(locale) {
+    this.intl.setLocale([locale, DEFAULT_LOCALE]);
+    this.dayjs.setLocale(locale);
+    this.dayjs.self.locale(locale);
```
Question: 
Le `setLocale` d'`ember-dayjs` à la ligne d'avant ne suffit pas ? Pourquoi appeler la méthode locale de `dayjs` ?

lego-technix replied:
C'est une question récurrente, la ligne `this.dayjs.self.locale(locale);` était présente avant la PR, c'est pourquoi je l'ai gardée. Je ne sais pas pourquoi elle est nécessaire.

er-lim replied:
Je viens de voir les réponses de Francois et Xavier, à enlever du coup 😸

lego-technix replied:
FAIT. Le push qui vient d'être fait à l'instant inclut la suppression de `dayjs.self.locale(locale)` dans toutes les applis avec des commits _refactor_.

er-lim commented:
Test fonctionnel sur RA (chrome 111) ✅

reibecca commented:
Testé fonctionnellement en local et sur les RA sur : 

- Chrome ✅
- Safari ✅

er-lim commented on orga/app/services/session.js:
```diff
@@ -3,20 +3,26 @@ import SessionService from 'ember-simple-auth/services/session';
-  @service currentDomain;
-  @service locale;
+
+  _localeFromQueryParam;
```
Question : 
Je vois que cette variable globale n'est utilisée que dans la fonction `handleLocale`. 
Pourquoi ne pas juste en faire une variable locale ? Pour garder la dernière locale en mémoire ?

lego-technix replied:
C'est effectivement pour garder en mémoire, c'est très important pour la logique souhaitée. Il suffit de supprimer cette ligne et voir comment les tests plantent pour en comprendre l'intérêt.

igorissen reviewed (commented):
- Tests sur Chrome ✅

igorissen commented on certif/tests/unit/services/session_test.js:
**Question** : Pourquoi l'avoir renommé en `service` ? Cela devrait plutôt être `session` simplement, ou resté `sessionService`.

lego-technix replied:
Parce que dans le même test dans Pix Certif cette variable est appelée `service`. Donc :
1. Ça permet de faire des diffs et de la synchronisation de code plus facilement entre Pix Certif et Pix Orga
2. Ça permet de faire des copier-coller de code entre tests de différents services au sein d'une même application plus facilement puisque c'est la même structure pour tous les tests de service puisque la variable principale qu'on veut tester est toujours nommée `service` !

reibecca commented on orga/app/services/session.js:
```diff
@@ -25,35 +31,42 @@ export default class CurrentSessionService extends SessionService {
-    if (domain === domainFr) {
-      this._setLocale(defaultLocale);
+    if (isFranceDomain) {
+      this._setLocale(FRENCH_INTERNATIONAL_LOCALE);
```
**Question :** La variable `FRENCH_INTERNATIONAL_LOCALE` est utilisée pour le domaine international, du coup il faudrait plutôt passer la `DEFAULT_LOCALE` vu que dans ce cas on est sur le domaine français non ? (c'est plus pour le nomage vu que la valeur est la même)

lego-technix replied:
Si on décide de traiter ce cas particulier `isFranceDomain` c'est qu'on ne veut pas mettre une valeur par défaut 🙂 C'est pourquoi j'ai effectué ce renommage de la variable de `defaultLocale` vers `FRENCH_INTERNATIONAL_LOCALE`.

Par contre on pourrait renommer `DEFAULT_LOCALE` en `DEFAULT_LOCALE`, qu'en pensez-vous ?

reibecca replied:
Ah oui d'accord je vois ! 🙂

Tu veux dire renommer `DEFAULT_LOCALE` en `FRENCH_INTERNATIONAL_LOCALE` ? Pour pouvoir le changer aussi dans la méthode privée `_setLocale()` ?

lego-technix replied:
Oui, c'est exactement à ça que je pense.

lego-technix replied:
FAIT. Le push qui vient d'être fait à l'instant inclut le renommage `DEFAULT_LOCALE`→`FRENCH_INTERNATIONAL_LOCALE`

reibecca commented on orga/mirage/config.js:
```diff
@@ -77,6 +78,12 @@ export default function () {
     return user;
   });
 
+  this.patch('/users/:id/lang/:lang', (schema, request) => {
+    const user = schema.users.find(request.params.id);
+    user.update({ lang: request.params.lang });
+    return user;
+  });
+
```
question : pourquoi avoir ajouté ce bout de code ? (c'est pour le test d'acceptance mais je ne me souviens plus comment ça fonctionne)

lego-technix replied:
Oui, c'est bien pour les tests d'acceptance 👍

Ce code permet de gérer le fait que l'appli Ember va faire un appel réseau pour modifier la `lang` du user quand on passe un `localeFromQueryParam`. Si on ne met pas ce bout de code, le test d'acceptance ne fonctionnera pas.

reibecca replied:
OK merci beaucoup 🙏🏼

annemarie35 commented on orga/app/controllers/authenticated/certifications.js:
```diff
@@ -4,12 +4,13 @@ import { action } from '@ember/object';
 
 export default class AuthenticatedCertificationsController extends Controller {
+  @service currentDomain;
+  @service url;
```
Le service n'est plus l'air utilisé, tu peux l'enlever

lego-technix replied:
Suppression du `@service url` qui n'est plus utilisé FAITE !

lego-technix committed:
refactor(certif): remove useless this.dayjs.self.locale(locale)

lego-technix committed:
refactor(orga): remove useless this.dayjs.self.locale(locale)

lego-technix committed:
refactor(mon-pix): remove useless this.dayjs.self.locale(locale)

lego-technix committed:
refactor(admin): remove useless this.dayjs.self.locale(locale)
